### PR TITLE
Prevent KVM from performing volume migrations of running instances

### PR DESCRIPTION
**Pull request by GutoVeronezi**
### Description
KVM has a limitation to modify instances definitions while they are on running state. Therefore, it is not possible to change volumes backend location easily.

There is a problem in the `migrateVolume` API. This API command ignores that limitation and causes an inconsistence on the database. ACS processes the migrate command, copies the volume to the destination storage, modifies the database and finishes the process with success. However, the running backend is still using the "old volume file".

This PR intends to prevent KVM to perform volumes migrations while KVM instances are in the running state and inform the user of an alternative API command that enables such operation on running instances.

### Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

#### Feature/Enhancement Scale

- [x] Major
- [ ] Minor

#### Bug Severity

- [ ] BLOCKER
- [ ] Critical
- [x] Major
- [ ] Minor
- [ ] Trivial

### How Has This Been Tested?
It has been tested locally.

### CODE REVIEW

**Review by GabrielBrascher — APPROVED** [review]
Code LGTM.

**Review by DaanHoogland — APPROVED** [review]
code lgtm

**Comment by rhtyd** [comment]
@blueorangutan package

**Review by weizhouapache — APPROVED** [review]
code makes sense.

could you please create pr for 4.15 instead of master ? (some changes on pom.xml needs to be reverted)

**Comment by rhtyd** [comment]
The PR base branch is 4.15 @weizhouapache

**Comment by weizhouapache** [comment]
> The PR base branch is 4.15 @weizhouapache, I think you mean rebase against 4.15

@rhtyd 
as I see, this PR is based on 4.15, but it have some commits from latest master (contains version update to 4.16.0.0) which are not needed.  should remove or revert these commits.

by the way, is there same issue in 4.14 ?

**Review by RodrigoDLopez — APPROVED** [review]
code LGTM.
great job @GutoVeronezi

**Comment by GutoVeronezi** [comment]
@RodrigoDLopez it's a good way to test it.

Before the commit, what you described in step 5 is what was happening.

To test it...

1. I created two primary storages.

2. I created one instance (using KVM as hypervisor) in one primary storage.

3. I updated hypervisor_capabilities to support storage motion of  a KVM.

4. Via GUI, I tried to migrate the volume to the other primary storage.

5. Before the modifications, it would claim that the volume was migrated, but actually it would create an inconsistency in the database and in the storage pool. Now, it's throwing an exception explaining why it is not possible and how to get around the situation.

**Comment by weizhouapache** [comment]
@GutoVeronezi @rhtyd 
can we have this in 4.14, 4.15 and master ?

**Comment by GutoVeronezi** [comment]
Thanks for the review guys. To us, it is fine if it only goes to master.

**Comment by weizhouapache** [comment]
> Thanks for the review guys. To us, it is fine if it only goes to master.

@GutoVeronezi 
the issue exists in 4.14 LTS and coming 4.15 LTS, we should fix it in the branches as well.
please update this pr

**Comment by GutoVeronezi** [comment]
> > Thanks for the review guys. To us, it is fine if it only goes to master.
> 
> @GutoVeronezi
> the issue exists in 4.14 LTS and coming 4.15 LTS, we should fix it in the branches as well.
> please update this pr

@weizhouapache 
We would rather not open for 4.14, it would be more work on our side. We already fixed it in our 4.13 internal branch, and now we are proposing the patch to contribute back with the community. Committers could execute a back port (if they think it is interesting) after this one is merged in master.

**Comment by weizhouapache** [comment]
> > > Thanks for the review guys. To us, it is fine if it only goes to master.
> > 
> > 
> > @GutoVeronezi
> > the issue exists in 4.14 LTS and coming 4.15 LTS, we should fix it in the branches as well.
> > please update this pr
> 
> @weizhouapache
> We would rather not open for 4.14, it would be more work on our side. We already fixed it in our 4.13 internal branch, and now we are proposing the patch to contribute back with the community. Committers could execute a back port (if they think it is interesting) after this one is merged in master.

@GutoVeronezi when you create a pr, it will trigger jenkins test, travis test, trillian test, and also be reviewed and/or manual tested by community. no matter which dest branch is.

**Comment by GutoVeronezi** [comment]
> > > > Thanks for the review guys. To us, it is fine if it only goes to master.
> > > 
> > > 
> > > @GutoVeronezi
> > > the issue exists in 4.14 LTS and coming 4.15 LTS, we should fix it in the branches as well.
> > > please update this pr
> > 
> > 
> > @weizhouapache
> > We would rather not open for 4.14, it would be more work on our side. We already fixed it in our 4.13 internal branch, and now we are proposing the patch to contribute back with the community. Committers could execute a back port (if they think it is interesting) after this one is merged in master.
> 
> @GutoVeronezi when you create a pr, it will trigger jenkins test, travis test, trillian test, and also be reviewed and/or manual tested by community. no matter which dest branch is.

@weizhouapache 
Ok, I see how it works. Anyways, to us it is fine it if it goes only to master.

**Comment by weizhouapache** [comment]
@GutoVeronezi ok, got it.
@DaanHoogland what's your opinion ? merge it into master then backport to 4.14/4.15 ?

**Review by shwstppr — APPROVED** [review]
LGTM

@weizhouapache @DaanHoogland @rhtyd do we plan to backport this or should this go into 4.15.1.0?

**Comment by DaanHoogland** [comment]
@weizhouapache @shwstppr @GutoVeronezi @rhtyd I am always for merging in the oldest applicable branch first. If we want this in those branches we should rebase it first (imnsho).

**Comment by weizhouapache** [comment]
> @weizhouapache @shwstppr @GutoVeronezi @rhtyd I am always for merging in the oldest applicable branch first. If we want this in those branches we should rebase it first (imnsho).

@DaanHoogland I agree. merge it to 4.14 at first, then merge to 4.15/master.

@GutoVeronezi can you create this branch from 4.14, apply the code change and force-push to github ?

**Comment by GutoVeronezi** [comment]
@weizhouapache done

**Review by weizhouapache — APPROVED** [review]
code lgtm

**Comment by weizhouapache** [comment]
> @weizhouapache done

@GutoVeronezi good, thanks

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
not sure if we need more third party testing on this. @rhtyd @weizhouapache, can we ?

**Comment by rhtyd** [comment]
Simple fix, I add a space between `if (...`. No testing needed, we don't support live volume migration with VM running.

**Pull request merged**
